### PR TITLE
feat: add .azdignore support for template init

### LINKED ISSUE #4142: [Issue] azd init should ignore some files
When a user runs `azd init` it copies all files from the repo.  Sometimes we don't want that.  

The template author may want a CONTRIBUTING.md file in the repo or a file in the .github folder for the template authoring experience.  But doesn't want the user to have that file when they are using the template.

Examples of files that we don't want to be copied when initialization a template, but should be in the repo.
https://github.com/Azure-Samples/agent-openai-python-prompty/blob/main/CONTRIBUTING.md
https://github.com/Azure-Samples/agent-openai-python-prompty/blob/main/CHANGELOG.md
https://github.com/Azure-Samples/agent-openai-python-prompty/blob/main/.github/CODE_OF_CONDUCT.md
https://github.com/Azure-Samples/agent-openai-python-prompty/blob/main/.github/workflows/azure-dev-validation.yaml

Suggestion:

Add a `.azdignore` file to the root of the repo that contains files/paths that azd should ignore during the init.

### PR DESCRIPTION
**Pull request by jongio**
## Add `.azdignore` support for `azd init`

Fixes #4142. Addresses #7669.

Revives the stale PR #4146 (by @jongio, 2024) with a clean reimplementation that incorporates all historical review feedback from #4146, the related packaging-ignore work (#4258, #5383), and multi-model expert code review.

### Problem

When running `azd init --template <repo>`, ALL files from the template repository are copied into the consumer's project — including contributor-only files like `SECURITY.md`, `CODE_OF_CONDUCT.md`, `.github/` CI configs, etc. This clutters consumer projects with files they don't need and didn't ask for (#7669).

### Solution

Template authors can now place a `.azdignore` file at the root of their template repository to exclude files from being copied when consumers run `azd init`. The file uses standard `.gitignore` syntax.

```
# .azdignore — files template consumers don't need
SECURITY.md
CODE_OF_CONDUCT.md
CONTRIBUTING.md
.github/
docs/
*.test.js
```

### History & Related Work

This feature has a long history across multiple issues and PRs spanning 2+ years:

| Reference | Type | Author | Status | Relationship |
|-----------|------|--------|--------|-------------|
| #4142 | Issue | @jongio | Open | Original `.azdignore` proposal (2024) |
| #7669 | Issue | @therealjohn | Open | Trigger issue — unwanted files in `azd init` |
| #4146 | PR | @jongio | Closed (stale) | First `.azdignore` implementation attempt |
| #4258 | PR | @jongio | Closed (superseded) | Packaging ignore — led to `.funcignore`/`.webappignore` |
| #5383 | PR | @weikanglim | Merged (1.18.0) | Shipped packaging ignore (`.funcignore`/`.webappignore`) |
| #1039 | Issue | @jongio | Closed | Early ignore discussion |
| #6754 | Issue | @therealjohn | Open | Related — preview files before init |

### Relationship to Existing Ignore Features

There are **two distinct ignore features** in azd — this PR implements only the first:

| Feature | File | Phase | Status |
|---------|------|-------|--------|
| **Init-time ignore** | `.azdignore` | `azd init` | **This PR** |
| Package-time ignore | `.funcignore` / `.webappignore` | `azd deploy` | Shipped 1.18.0 (#5383) |
| Docker ignore | `.dockerignore` | Docker build | Already supported |
| Git ignore | `.gitignore` | `azd init` (local copy) | Already supported |

All use the same `go-gitignore` library and root-only pattern.

### Design Decisions

Every design choice was informed by review feedback from the original PRs:

| Decision | Rationale | Feedback Source |
|----------|-----------|----------------|
| **Separate `.azdignore` file** (not azure.yaml) | `.gitignore` syntax is universally understood; YAML has glob limitations | @pamelafox #4258 |
| **Root-only** (no recursive nested files) | Init copies entire template; nested adds complexity with minimal benefit | @weikanglim #4258 |
| **Self-excluding** (`.azdignore` removed from output) | Consumers don't need template authoring metadata | — |
| **Internal helpers** (no shared package) | Avoids premature abstraction; keeps code scoped to init | @weikanglim #4258 |
| **`removeAzdIgnoredFiles(dir)` loads file internally** | Cleaner API than passing matchers around | @wbreza #4146 |
| **All helpers unexported** | Only used within repository package | @wbreza #4146 |
| **UTF-8 BOM stripping** | Windows editors prepend BOM that breaks first pattern | Pattern from #5383 |
| **Symlink `.azdignore` rejected** | Prevents reading arbitrary files outside staging | Expert review finding |
| **`.azdignore` survives `.gitignore` filtering** | Local templates with broad `.gitignore` patterns must not shadow `.azdignore` | Expert review finding |
| **1MB size limit with `io.LimitReader`** | DoS prevention; TOCTOU-safe via atomic read | Expert review finding |

### Historical Review Feedback — Full Resolution Table

All 23 direct-relevance comments from original PRs/issues tracked and resolved:

<details>
<summary>PR #4146 review comments (13 items)</summary>

| Reviewer | Feedback | Resolution |
|----------|----------|------------|
| @wbreza | Revert launch.json changes | N/A — no launch.json changes |
| @ellismg | Nervous about local path template support, Windows path munging | Removed — no path.go changes |
| @weikanglim | Local path could allow remote sources to reference local paths (security) | Not included |
| @ellismg | ssh:// template support untested | Not included |
| @ellismg | "Wonderful test!" on functional init test | ✅ |
| @weikanglim | `filepath.Separator` on Windows breaks trailing slash trim | N/A — no path.go changes |
| @wbreza | `RemoveIgnoredFiles` should internally load ignore file | ✅ `removeAzdIgnoredFiles(dir)` calls `loadAzdIgnore` internally |
| @wbreza | Rename `staging` to `stagingPath` | ✅ Uses `dir` parameter name |
| @wbreza | Don't export `LoadIgnoreFiles` if internal | ✅ `loadAzdIgnore` is unexported |
| @wbreza | Don't export other helper functions | ✅ All helpers unexported |
| @wbreza | Did we land on `.zipignore` naming? | N/A — packaging uses `.funcignore`/`.webappignore` (#5383) |
| @wbreza | Test function naming: `Test_FunctionName` | ✅ Tests follow `Test_functionName` convention |
| @wbreza | Move .azdignore test sample to file | Tests use inline content for clarity and test isolation |

</details>

<details>
<summary>PR #4258 review comments (5 items)</summary>

| Reviewer | Feedback | Resolution |
|----------|----------|------------|
| @pamelafox | Why separate file vs azure.yaml `--ignore`? | Separate file: gitignore syntax is standard |
| @weikanglim | Don't create shared `dotignore` package prematurely | ✅ Helpers in `initializer.go`, no separate package |
| @weikanglim | Should we support recursive/nested ignore files? | Root-only, matching packaging approach |
| @weikanglim | gitignore-spec conformance concerns | Uses `go-gitignore` (same lib as packaging) |
| @weikanglim | Separate rzip changes from ignore changes | N/A — no rzip changes |

</details>

<details>
<summary>Issue #4142 comments (5 items)</summary>

| Author | Feedback | Resolution |
|--------|----------|------------|
| @jongio | Proposed `.azdignore` file using gitignore syntax | Implemented as proposed |
| @vhvb1989 | Suggested `azd-template` branch convention | Rejected: requires separate branch maintenance |
| @wbreza | Supported `.azdignore` approach | ✅ |
| @jongio | Clarified init-time vs packaging-time distinction | Maintained throughout |
| @jongio | Rejected branch convention — error prone | ✅ |

</details>

#### Integration Point

```
azd init --template <repo>
  │
  ├── Remote template (git clone → staging dir)
  │     └── removeAzdIgnoredFiles(staging)
  │
  └── Local template (copyLocalTemplate → staging dir)
        ├── .azdignore preserved through .gitignore filtering
        └── removeAzdIgnoredFiles(staging)
```

Both paths converge on `removeAzdIgnoredFiles(staging)` — a single code path for uniform behavior.

#### Security Hardening

- **Symlink rejection**: `os.Lstat` rejects symlink/non-regular `.azdignore` files
- **TOCTOU-safe size limit**: `io.LimitReader` enforces 1MB on actual bytes read
- **Path traversal safe**: `filepath.WalkDir` constrains to staging directory
- **`.gitignore` bypass protection**: `.azdignore` is never filtered by `.gitignore` during local copy

#### Files Changed

| File | Lines | Description |
|------|-------|-------------|
| `cli/azd/internal/repository/initializer.go` | +137 | Core `.azdignore` logic |
| `cli/azd/internal/repository/initializer_test.go` | +773 | 23 test cases |
| `cli/azd/.vscode/cspell.yaml` | +1 | Added "azdignore" to dictionary |

#### Test Coverage (23 cases)

| Category | Tests |
|----------|-------|
| **Unit** | `removeAzdIgnoredFiles` (5 patterns), `loadAzdIgnore` (3 cases), `filterExistingFiles` |
| **Integration** | Remote template, local template, negation patterns, .gitignore coexistence, .gitignore survival |
| **Security** | Path traversal, symlink rejection, empty/comment-only, malformed content, negation-only patterns, Unicode filenames, symlinks skipped, source-is-symlink, large pattern file |

### Expert Review Summary

This implementation was reviewed by multiple independent expert models:

| Focus | Findings | Status |
|-------|----------|--------|
| azd architecture | No issues | ✅ |
| Go idioms & security | 1 HIGH: symlink `.azdignore` reads arbitrary files | ✅ Fixed |
| Correctness | 1 MEDIUM: `.gitignore` can shadow `.azdignore` in local templates | ✅ Fixed |

### CODE REVIEW

File: cli/azd/internal/repository/initializer_test.go [hunk 1]
```diff
@@ -1182,3 +1183,775 @@ func Test_Initializer_Initialize_LocalTemplateOverlapRejected(t *testing.T) {
...
+	sb.WriteString("*.log\n")
+	require.NoError(t, os.WriteFile(
+		filepath.Join(dir, azdIgnoreFileName), []byte(sb.String()), 0600))
+
+	err := removeAzdIgnoredFiles(dir)
+	require.NoError(t, err)
+
+	require.FileExists(t, filepath.Join(dir, "keep.txt"))
+	require.NoFileExists(t, filepath.Join(dir, "remove-me.log"))
+}
```
Comment (spboyer) [inline cli/azd/internal/repository/initializer_test.go:2027]: **[LOW]** Testing: `loadAzdIgnore` has two size enforcement points — the `info.Size()` Lstat check and the post-read `LimitReader` TOCTOU double-check — but neither path is exercised by any test. `Test_removeAzdIgnoredFiles_LargePatternFile` creates ~200KB of content, well under the 1MB boundary.

Consider adding a boundary test:
```go
func Test_loadAzdIgnore_RejectsOversizedFile(t *testing.T) {
    t.Parallel()
    dir := t.TempDir()
    data := bytes.Repeat([]byte("x"), azdIgnoreMaxSize+1)
    require.NoError(t, os.WriteFile(
        filepath.Join(dir, azdIgnoreFileName), data, 0600))
    ig, err := loadAzdIgnore(dir)
    require.Error(t, err)
    require.Nil(t, ig)
    require.Contains(t, err.Error(), "exceeds maximum size")
}
```

**Review by spboyer — COMMENTED** [review]
Well-structured implementation with thorough security hardening and comprehensive test coverage (23 cases). The design decisions are well-documented and the code follows all repo conventions. One minor test gap noted.

- initializer_test.go — missing boundary test for the 1MB `azdIgnoreMaxSize` enforcement in `loadAzdIgnore`

**Review by spboyer — APPROVED** [review]
LGTM — clean implementation with thorough security hardening and excellent test coverage. One minor test gap noted in prior comment.

File: cli/azd/internal/repository/initializer_test.go (hunk 1 above)
Comment (jongio) [reply cli/azd/internal/repository/initializer_test.go:2027]: Good catch -- added `Test_loadAzdIgnore_RejectsOversizedFile` in 5415a4eb4. Creates a file at `azdIgnoreMaxSize+1` bytes and verifies the `exceeds maximum size` error from the `info.Size()` Lstat check path.

**Review by wbreza — COMMENTED** [review]
## Code Review — PR #7685: .azdignore support for template init

**Verdict: 💬 Comments/Suggestions** — Code is solid with strong security hardening. A few gaps worth addressing.

### Findings

#### 🟠 1. No user-facing documentation
Template authors have no way to discover this feature. There's no authoring guide, help text, or pattern syntax docs. At minimum, a brief section in template authoring docs would help:
- What \.azdignore\ is and where to place it
- Supported syntax (standard \.gitignore\ patterns)
- Examples (excluding \.github/\, CI files, contributor docs)
- Behavior note: \.azdignore\ itself is removed from consumer projects

**Suggestion:** Add docs pre-merge or as a committed fast-follow.

#### 🟡 2. Nested \.azdignore\ files not cleaned up
\
emoveAzdIgnoredFiles()\ only removes the root \.azdignore\. If a template has nested \.azdignore\ files (e.g., \docs/.azdignore\), they leak into the consumer's project. This violates the self-excluding guarantee.

**Suggestion:** Walk the staging dir and remove ALL files named \.azdignore\, not just the root one. ~5-line fix.

#### 🟡 3. Missing test for \**\ recursive glob pattern
The \**\ pattern (e.g., \**/node_modules\) is the most common \.gitignore\ pattern users will try. While \denormal/go-gitignore\ likely supports it, there's no test confirming this works through the \.azdignore\ integration.

**Suggestion:** Add one table-driven test entry with a \**\ pattern to verify recursive matching works end-to-end.

### What's Good
- ✅ Strong security: symlink rejection, TOCTOU-safe size limits, path traversal protection
- ✅ UTF-8 BOM handling for Windows editors
- ✅ 23 test cases covering edge cases (traversal, malformed content, Unicode, oversized files)
- ✅ Defensive regression test ensuring \.azdignore\ survives broad \.gitignore\ patterns like \.*\
- ✅ Clean integration for both remote and local template paths

Nice work on the security hardening especially — the \Lstat\ + \LimitReader\ combo is solid. 👍

**Comment by jongio** [comment]
Thanks @wbreza — all three addressed in e7e34c1be:

1. **Nested .azdignore cleanup**: `removeAzdIgnoredFiles()` now walks the staging dir and removes ALL `.azdignore` files at any depth, not just root. Added `Test_removeAzdIgnoredFiles_NestedAzdIgnoreCleanup` to verify.

2. **`**` recursive glob test**: Added `RecursiveDoubleStarPattern` entry that creates `node_modules/` at 3 depths and verifies `**/node_modules` excludes all of them.

3. **Documentation**: Added `cli/azd/docs/azdignore.md` covering placement, syntax, examples, behavior, and security notes.

**Review by wbreza — APPROVED** [review]
## Re-Review — PR #7685: .azdignore support

**Verdict: ✅ Approve**

All 3 suggestions from my earlier review have been addressed in the latest commit:

1. ✅ **User-facing documentation** — New \docs/azdignore.md\ with syntax reference, examples, placement guidance, and security notes
2. ✅ **Nested .azdignore cleanup** — Second \WalkDir\ pass removes ALL \.azdignore\ files at any depth, with dedicated test (\Test_removeAzdIgnoredFiles_NestedAzdIgnoreCleanup\)
3. ✅ **\**\ recursive glob test** — \RecursiveDoubleStarPattern\ test verifies matching at root, 1-deep, and 2-deep directories

No new issues introduced. Clean implementation with strong security hardening (symlink rejection, path traversal protection, size limits, BOM handling). Ship it! 🚀
